File: platform/fastify_server/routes/csp_report.route.mjs
/**
 * CSP Violation Report Endpoint
 *
 * Receives browser Content-Security-Policy violation reports in both
 * legacy (application/csp-report) and modern (application/reports+json)
 * Reporting API formats.
 */

/** @type {ReadonlyArray<string>} Fields to extract from violation reports */
const REPORT_FIELDS = [
  'document-uri',
  'violated-directive',
  'effective-directive',
  'blocked-uri',
  'script-sample',
  'disposition',
  'source-file',
  'line-number',
  'column-number',
];

/**
 * Normalize a single violation report from either legacy or modern format.
 * @param {object} body - Raw report body
 * @returns {object} Normalized violation fields
 */
function normalizeReport(body) {
  const report = {};
  for (const field of REPORT_FIELDS) {
    if (body[field] !== undefined) {
      report[field] = body[field];
    }
  }
  return report;
}

/**
 * Mount CSP report routes to the Fastify application.
 * @param {import('fastify').FastifyInstance} server
 */
export async function mount(server) {
  // Register content-type parsers for CSP report formats
  server.addContentTypeParser('application/csp-report', { parseAs: 'string' }, (_req, body, done) => {
    try {
      done(null, JSON.parse(body));
    } catch (err) {
      done(err);
    }
  });

  server.addContentTypeParser('application/reports+json', { parseAs: 'string' }, (_req, body, done) => {
    try {
      done(null, JSON.parse(body));
    } catch (err) {
      done(err);
    }
  });

  server.post('/api/csp-report', async (request, reply) => {
    const contentType = request.headers['content-type'] || '';
    const body = request.body;

    /** @type {object[]} */
    let violations = [];

    if (contentType.includes('application/csp-report') && body?.['csp-report']) {
      // Legacy format: { "csp-report": { ... } }
      violations = [normalizeReport(body['csp-report'])];
    } else if (contentType.includes('application/reports+json') && Array.isArray(body)) {
      // Modern Reporting API format: [{ type, body: { ... } }, ...]
      violations = body
        .filter((entry) => entry.type === 'csp-violation' && entry.body)
        .map((entry) => normalizeReport(entry.body));
    } else if (body && typeof body === 'object') {
      // Fallback: try to extract what we can
      violations = [normalizeReport(body['csp-report'] || body)];
    }

    for (const violation of violations) {
      server.log.warn({ cspViolation: violation }, 'CSP violation report received');
    }

    reply.code(204);
  });
}

File: platform/fastapi_server/routes/csp_report.route.py
"""
CSP Violation Report Endpoint

Receives browser Content-Security-Policy violation reports in both
legacy (application/csp-report) and modern (application/reports+json)
Reporting API formats.
"""

import json
import logging

from fastapi import FastAPI, Request

logger = logging.getLogger("csp_report")

REPORT_FIELDS = (
    "document-uri",
    "violated-directive",
    "effective-directive",
    "blocked-uri",
    "script-sample",
    "disposition",
    "source-file",
    "line-number",
    "column-number",
)


def _normalize_report(body: dict) -> dict:
    """Extract known violation fields from a report body."""
    return {field: body[field] for field in REPORT_FIELDS if field in body}


def mount(app: FastAPI):
    """Mount CSP report routes to the FastAPI application."""

    @app.post("/api/csp-report", status_code=204)
    async def csp_report(request: Request):
        raw = await request.body()
        if not raw:
            return

        try:
            data = json.loads(raw)
        except (json.JSONDecodeError, UnicodeDecodeError):
            logger.warning("CSP report: invalid JSON body")
            return

        content_type = request.headers.get("content-type", "")
        violations: list[dict] = []

        if "application/csp-report" in content_type and isinstance(data, dict):
            # Legacy format: { "csp-report": { ... } }
            csp_report_body = data.get("csp-report", data)
            violations = [_normalize_report(csp_report_body)]
        elif "application/reports+json" in content_type and isinstance(data, list):
            # Modern Reporting API format: [{ type, body: { ... } }, ...]
            violations = [
                _normalize_report(entry["body"])
                for entry in data
                if entry.get("type") == "csp-violation" and isinstance(entry.get("body"), dict)
            ]
        elif isinstance(data, dict):
            # Fallback: try to extract what we can
            csp_report_body = data.get("csp-report", data)
            violations = [_normalize_report(csp_report_body)]

        for violation in violations:
            logger.warning("CSP violation report received: %s", violation)
